k4Reco/DDPlanarDigi/components/DDPlanarDigi.cpp

```diff
@@ -0,0 +1,348 @@
+                           KeyValue("HeaderName", "EventHeader"),
+                       },
+                       {KeyValue("TrackerHitCollectionName", "VTXTrackerHits"),
+                        KeyValue("SimTrkHitRelCollection", "VTXTrackerHitRelations")}) {
+  m_uidSvc = service<IUniqueIDGenSvc>("UniqueIDGenSvc", true);
+  if (!m_uidSvc) {
+    error() << "Unable to get UniqueIDGenSvc" << endmsg;
+  }
+
+  if (m_resULayer.size() != m_resVLayer.size()) {
+    error() << "DDPlanarDigi - Inconsistent number of resolutions given for U and V coordinate: "
+            << "ResolutionU  :" << m_resULayer.size() << " != ResolutionV : " << m_resVLayer.size();
+
+    throw std::runtime_error("DDPlanarDigi: Inconsistent number of resolutions given for U and V coordinate");
+  }
+
+  m_histograms.resize(hSize);
+
+  m_histograms[hu].reset(new Gaudi::Accumulators::Histogram<1>{this, "hu", "smearing u", {50, -5., +5.}});
+  m_histograms[hv].reset(new Gaudi::Accumulators::Histogram<1>{this, "hv", "smearing v", {50, -5., +5.}});
+  m_histograms[hT].reset(new Gaudi::Accumulators::Histogram<1>{this, "hT", "smearing time", {50, -5., +5.}});
+
+  m_histograms[diffu].reset(new Gaudi::Accumulators::Histogram<1>{this, "diffu", "diff u", {1000, -.1, +.1}});
+  m_histograms[diffv].reset(new Gaudi::Accumulators::Histogram<1>{this, "diffv", "diff v", {1000, -.1, +.1}});
+  m_histograms[diffT].reset(new Gaudi::Accumulators::Histogram<1>{this, "diffT", "diff time", {1000, -5., +5.}});
+
+  m_histograms[hitE].reset(new Gaudi::Accumulators::Histogram<1>{this, "hitE", "hitEnergy in keV", {1000, 0, 200}});
+  m_histograms[hitsAccepted].reset(
+      new Gaudi::Accumulators::Histogram<1>{this, "hitsAccepted", "Fraction of accepted hits [%]", {201, 0, 100.5}});
+
+  m_geoSvc = serviceLocator()->service(m_geoSvcName);
+}
+
+StatusCode DDPlanarDigi::initialize() {
+  const auto detector = m_geoSvc->getDetector();
+
+  const auto         surfMan = detector->extension<dd4hep::rec::SurfaceManager>();
+  dd4hep::DetElement det     = detector->detector(m_subDetName.value());
+  surfaceMap                 = surfMan->map(m_subDetName.value());
+
+  if (!surfaceMap) {
+    std::stringstream err;
+    err << " Could not find surface map for detector: " << det.name() << " in SurfaceManager ";
+    throw std::runtime_error(err.str());
+  }
+
+  // Get and store the name for a debug message later
+  (void)this->getProperty("SimTrackerHitCollectionName", m_collName);
+
+  return StatusCode::SUCCESS;
+}
+
+std::tuple<TrackerHitPlaneColl, Association> DDPlanarDigi::operator()(
+    const SimTrackerHitCollection& simTrackerHits, const Header& headers) const {
+  auto seed = m_uidSvc->getUniqueID(headers[0].getEventNumber(), headers[0].getRunNumber(), this->name());
+  info() << "Using seed " << seed << " for event " << headers[0].getEventNumber() << " and run "
+         << headers[0].getRunNumber() << endmsg;
+  m_engine.seed(seed);
```

Comment:
What is the size of the `seed` in this case? The mersenne twister needs a comparatively large seed to properly initialize it's internal state. See, e.g. https://codereview.stackexchange.com/a/109266

I am not entirely sure that we have to do this for all events, but maybe we should initialize it properly (and reproducibly) once in `initialize`.

Reply:
The seed is `uint64_t`. But, as long as the seeds for every event are different then it shouldn't matter which seeds are being used . In the link I only see that if you have 2^32 seeds you have at most 2^32 choices for the first generated number so of course you can find them. For all events or not is a different discussion. Even when compiling without optimizations seeding a million times takes a few seconds so I don't think it's much of a problem. The idea of the `UniqueIDGenSvc` was certainly to do this for each event and in the original processor it was being done for each event.

Reply:
The problem with an improperly seeded MT (as far as I understand it), is that it will not be possible to reach all internal states (and thus random outputs?). So the actual amount of unique random numbers that you get might be much smaller than what is theoretically possible. I am mainly raising this because the Marlin processor used the GSL random library, specifically [`gsl_rng_ranlxs2`](https://www.gnu.org/software/gsl/doc/html/rng.html#c.gsl_rng_ranlxs2), which seems to be "more random" than the MT. I have://www.gnu.org/software/gsl/doc/html/rng.html#c.gsl_rng_ranlxs2), which seems to be "more random" than the MT. I am not sure any of this matters in the end, but it might be the source of somewhat subtle bugs since the random engine changes from Marlin to Gaudi.

Reply:
@jmcarcell  @tmadlener choosing a random engine is not easy... Geant4 default random engine is the so-called MinMax, described [here](https://www.sciencedirect.com/science/article/pii/S0010465515002489?via%3Dihub), where they also comment about the MT the following:

> In this sense, the underlying dynamical system of the Mersenne Twister has one order of magnitude less entropy than RCARRY. This is related to the singular flaw acknowledged by the authors of the Mersenne Twister, which is that MT has a very long recovery time when it is seeded by a vector containing mostly zeros: the output is observed to be non-random even after outputting a million values. In this instance, the divergence of some trajectory is observed away from the origin. In fact, from the dynamical system point of view this is not merely a manifestation of an unlucky initialization, since any two nearby trajectories of the MT diverge very slowly and this is ultimately what causes the failure of MT in the statistical tests. The total entropy of the MT system is approximately h ≈ 4.8. On the basis of the known behaviour of RCARRY [10,11] and our investigation of the MIXMAX, it appears that an entropy of at least h & 50 is required for a generator to have a sufficiently random trajectory. In light of this, it is perhaps not surprising that the Mersenne Twister fails many tests in its pure form, and still fails some tests even when some additional tempering of the output is applied

Reply:
@tmadlener why not using the [ROOT wrapper](https://root.cern.ch/doc/master/classROOT_1_1Math_1_1GSLRngRanLuxD2.html#ab67a7d9bf0e41d317d82112fdc95a3ad) for the GSL random engine?

Reply:
> @tmadlener why not using the [ROOT wrapper](https://root.cern.ch/doc/master/classROOT_1_1Math_1_1GSLRngRanLuxD2.html#ab67a7d9bf0e41d317d82112fdc95a3ad) for the GSL random engine? Edit, I just found a bug that prevents its use at the moment...

Most likely because this is a port from the corresponding Marlin processor that did not have a ROOT dependency.

Reply:
Oki, fair enough. In case you want to avoid depending on GSL random engines (because of [software license](https://janelia-flyem.github.io/licenses.html)), boost offers the same engines, [link](https://www.boost.org/doc/libs/1_82_0/doc/html/boost_random/reference.html#boost_random.reference.generators).

Reply:
This is using now `TRandom2`. I will merge this shortly since there hasn't been much discussion other than the random number generator, and the validation should already be done, although I will try to run again.